Enable stable edge ID-based custom speeds for autos

mark-idleman opened the pull request:
## Background

Follows up on #180, which moved our setting of SEIDs earlier in the OSM import process, to enable custom speeds at a SEID granularity. The existing OSM-based custom speed functionality is preserved, with new generic classes introduced to represent custom speeds more generally that can hold either OSM or SEID speeds.

Given the size of the changes, and some mode-based idiosyncrasies that my approach required, I decided to only enable the feature for autos in this PR; a follow-up PR will enable SIED custom speeds for all other modes.

## Overview

The changes here generally fall into two categories: (1) updating the existing places where custom speeds are handled to instead use generic classes that can hold either custom speed type and (2) updating the spot where speeds are applied to new graphhopper edges to apply custom SEID-based speeds, if they exist.

- For new generic classes, we now have `CustomSpeed` to store the actual custom speed values (eg 50kph), and `CustomSpeedId` to store the OSM ID/SEID associated with those speed values. These classes are used in much of the initial parsing/passing around of custom speeds during the import process; they are only "unwrapped" into their type-specific subclasses (`OsmId`/`StableEdgeId` and `DirectedSpeeds`/`NetworkLinkSpeed`) when the speeds are being applied
- When we initially parse custom speeds in `CustomSpeedsUtils.parseIdToCustomSpeed()`, we now determine based on ID column name whether the speeds are SEIDs or OSM IDs, and create objects of the appropriate type to store those speeds/IDs. From this point on, speeds are stored using the generic parent classes (`CustomSpeed` and `CustomSpeedId`), and we unwrap them only when we need to access them
- When we validate the incoming custom speeds in `CustomSpeedsVehicle.validateCustomSpeeds()`, we mimic the existing OSM-based speed validation logic for SEIDs (if present)
- Finally, we update the tag parser and access parser classes for each mode to handle SEIDs
  - The tag parser classes (eg `ReplicaCustomSpeedsCarTagParser`) hold the `handleWayTags` function, which is where speeds that have been parsed from OSM are applied to internal graphhopper edges as they're created. The current OSM-based custom speed logic is called in this function (either directly, or in an overridden function like `applyMaxSpeed` that's called in `handleWayTags`). To allow for both types of custom speeds, we simply gate the new SEID-based logic added to `handleWayTags` and the existing OSM-based logic with `if` statements checking the custom speeds type; if the provided speeds are OSM-based, we won't run the SEID-based custom speed logic, and vice-versa
  - Note that having access to SEIDs, necessary to apply SEID-based custom speeds in `handleWayTags`, is only possible because of the changes we made in #180 to generate SEIDs earlier in `CustomOsmReader`
  - The access parser classes (eg `ReplicaCustomSpeedsCarAccessParser`) handle the accessibility of graphhopper edges by each mode, which is separate from the speeds stored for those edges. Here, we make sure that the directed OSM custom speed validation (`validateCustomSpeedDirection`) is only called when the custom speed type is OSM, and we update the `removeAccessForUserClosedRoads` function to handle both OSM and SEID custom speeds
  
For non-car modes, I check if SEID-based custom speeds are provided in `handleWayTags`, and if so, throw an error. Because this function is guaranteed to run for every (mode + edge) during the import process, it effectively guarantees that SEID-based custom speeds provided for non-auto modes are not allowed.
  
## Testing
- Unit tests added to cover all of the new functions/logic I added (I think!). This includes tests that cover new profiles I added to our test config that simulate a road closure (`car_custom_seid_closed_baseline_road`) and making a road's speeds faster (`car_custom_seid_fast_thurton_drive`)
- Graphhopper repo [functional test passes](https://github.com/replicahq/graphhopper/actions/runs/14763326904); this shows that exact routes (in a non-scenario context) should stay stable with this change, at least for auto
- We tested an earlier version of this branch to help @sudatta-mohanty test his recent changes to update the router builder DM to allow for SEID-based custom speeds; the road closures we tried worked as expected when tested in the GUI

## Individual closure testing
Several individual auto closures were tested locally. I kept one car profile unmodified, to show the difference between the auto routing without closures vs. with closures applied. I try to specify in each screenshot what changed in the results from the closure profile.

#### Eastbound closure of I-80 on the yolo bypass:
<img width="600" alt="Screenshot 2025-05-07 at 11 02 23 AM" src="https://github.com/user-attachments/assets/d2917148-0d28-4470-a5ad-32e70cf75cc9" />

We're able to route along I-80 heading west, but a detour is taken going east:
<img width="600" alt="Screenshot 2025-05-07 at 12 03 38 PM" src="https://github.com/user-attachments/assets/2c853ad8-4346-4fb3-92c9-6343b8b540ff" />
<img width="600" alt="Screenshot 2025-05-07 at 12 03 06 PM" src="https://github.com/user-attachments/assets/e8f71324-deb1-4ff3-8778-d24c6b04d8ba" />

#### On-ramp near Davis:
<img width="600" alt="Screenshot 2025-05-07 at 11 03 00 AM" src="https://github.com/user-attachments/assets/8c0cf441-3e8a-4ebd-a3b4-fd5b06763199" />

The on-ramp is ignored and instead a long detour to the next highway exit is taken:
<img width="600" alt="Screenshot 2025-05-07 at 12 07 04 PM" src="https://github.com/user-attachments/assets/239a2a29-761e-4ef3-999c-a7b1176b2958" />

#### Northbound closure of rural road east of Sac:
<img width="600" alt="Screenshot 2025-05-07 at 11 06 14 AM" src="https://github.com/user-attachments/assets/4b67ff23-d999-49ae-9c1e-19b96fb39f39" />

With closure applied, a clearly longer route is chosen:
<img width="600" alt="Screenshot 2025-05-07 at 12 12 34 PM" src="https://github.com/user-attachments/assets/315b991a-b90f-4638-88d4-ef86acecff27" />

#### One direction of bridge in Sac:
<img width="600" alt="Screenshot 2025-05-07 at 11 05 03 AM" src="https://github.com/user-attachments/assets/e0d00dfa-32bf-4e32-82ea-90163870b492" />

A spaghetti of alternatives is created by the closure routing profile:
<img width="600" alt="Screenshot 2025-05-07 at 12 21 20 PM" src="https://github.com/user-attachments/assets/e2f73da2-5827-40c4-8350-fe468e6c7a38" />

mark-idleman committed:
initial sketch comments

mark-idleman committed:
Merge branch 'original-direction' into seid_based_custom_speeds

mark-idleman committed:
first full pass

mark-idleman committed:
switch actions/cache to v4

mark-idleman committed:
fix generic to specific custom speed class casting

mark-idleman committed:
make sure test graph files are built before custom speed utils test

mark-idleman committed:
hopefully get tests passing

mark-idleman committed:
fix pt GUI

mark-idleman committed:
add first seid-based unit test

mark-idleman committed:
parse osm IDs prior to setting SEIDs during OSM parsing; add new SEID-based closure test

mark-idleman committed:
oops add CustomOSMParsers

mark-idleman committed:
fix router server test

mark-idleman committed:
remove post-import calculation of SEIDs

mark-idleman committed:
Merge branch 'original-direction' into seid_based_custom_speeds

mark-idleman committed:
(attempt) depend on replica graphhopper_main instead of officially published gh package

mark-idleman committed:
(attempt) depend on replica graphhopper_main instead of officially published gh package

mark-idleman committed:
try to fix build by installing gcloud manually in maven-build step

mark-idleman committed:
try again

mark-idleman committed:
try installing gcloud in yet another way

mark-idleman committed:
one more time

mark-idleman committed:
twice more time

mark-idleman committed:
new try

mark-idleman committed:
fix copy path

mark-idleman committed:
try adding new gcs repos to maven_settings

mark-idleman committed:
Revert "try adding new gcs repos to maven_settings"

This reverts commit e0f99e43ad7dbc05e444c7934acdaacd872e625d.

mark-idleman committed:
try bumping storage wagon version:

mark-idleman committed:
Try setting GOOGLE_APPLICATION_CREDENTIALS in mvn build step

mark-idleman committed:
Revert "try bumping storage wagon version:"

This reverts commit 53970a723939a6eccf4289ba481ec7b0374fae0b.

mark-idleman committed:
cleanup

mark-idleman committed:
remove GOOGLE_APPLICATION_CREDENTIALS env var

mark-idleman committed:
remove old gh core reference

mark-idleman committed:
Revert "remove GOOGLE_APPLICATION_CREDENTIALS env var"

This reverts commit 16601f75b1fea2cabaa2dccd08b11b33f4fca7d7.

mark-idleman committed:
try removing web-bundle dep from web pom

mark-idleman committed:
Revert "try removing web-bundle dep from web pom"

This reverts commit 4bd88a3ed2e33c6d04004252a1fadaed7bcf990c.

mark-idleman committed:
merge point_to_graphhopper_main

mark-idleman committed:
fix applying of foot + bike SEID-based custom speeds; add basic SEID closure test for all modes

mark-idleman committed:
add new custom speed file (SEID-based fast thurton drive)

mark-idleman committed:
fix testGetCustomSpeedVehiclesByName

mark-idleman committed:
cleanup unused function

mark-idleman committed:
merge original-direction

mark-idleman committed:
only store both directions for speeds, not priority, for bikes; reset affected unit tests to check for old values

mark-idleman committed:
cleanup

mark-idleman committed:
first pass at removing SEID closure support for non-auto modes

mark-idleman committed:
fix current tests (I think)

mark-idleman committed:
new test

mark-idleman committed:
add new seid-based test for removeAccessForUserClosedRoads()

mark-idleman committed:
add getCustomSpeedVehiclesByName test for fast thurton drive SEID custom speed file

mark-idleman committed:
more explicit long

mark-idleman committed:
a few more small CustomSpeedsUtils tests

mark-idleman committed:
new CustomSpeedsVehicleTest for seid speeds

mark-idleman commented on web-bundle/src/main/java/com/graphhopper/replica/ReplicaCustomSpeedsFootTagParser.java:
```diff
@@ -31,24 +37,31 @@ public void handleWayTags(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way
+            throw new NotImplementedException("SEID-based custom speeds are not yet implemented for foot!");
+        } else {
+            ImmutableMap<Long, DirectedSpeeds> osmWayIdToDirectedMaxSpeeds = getOsmBasedCustomSpeeds(idToCustomSpeeds);
+            double forwardSpeed = CustomSpeedsUtils.getCustomMaxSpeed(way, osmWayIdToDirectedMaxSpeeds, false).orElse(speed);
```
Note: we're refactoring a bit here to make the things cleaner (this function was always called with `fwd` and `bwd` set to true, so we could remove those params) but lines 60 + 61 are also fixing what was technically incorrect logic. 

Before, we were only grabbing the osm-based custom speed in the forward direction, and applying it to both directions of the edge if it was present; now, we grab the forward and backward custom speds, and apply them appropriately in each separate direction

mark-idleman committed:
update config comment

mark-idleman commented on web-bundle/src/main/java/com/graphhopper/replica/ReplicaCustomSpeedsBikeTagParser.java:
```diff
@@ -1,23 +1,43 @@
+
+    @Override
+    public void handleWayTags(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way) {
+        super.handleWayTags(edgeId, edgeIntAccess, way);
```
Here's an example of how we throw errors when SEID-based custom speeds have been provided for a non-supported mode, like bike. Note that OSM-based speeds are still handled as they were in `applyMaxSpeed` if they're provided, which is called within `super.handleWayTags()` here

kaelgreco replied:
do we need to apply these checks to other modes as well?

sudatta-mohanty reviewed (commented):
The change looks good to me and I tested an earlier version of the change for car closures already, so I'm confident that it works as expected! 

The only question I have is around the longer term plan for the additional code introduced here. There are a lot of additional checks now for `CustomSpeed` and `CustomSpeedId`. Assuming we eventually transition to only supporting SEID-based closures, is the plan to get rid of all these checks in the future? And would that be a lot of effort?

mark-idleman commented on web-bundle/src/main/java/com/graphhopper/customspeeds/CustomSpeedsUtils.java:
```diff
@@ -85,6 +87,57 @@ private static Map<String, File> getVehicleNameToCustomSpeedFile(List<Profile> p
+            for (CSVRecord record : parser) {
+                // If row has incorrect number of columns (indicating a jagged CSV), error
+                if (!record.isConsistent()) {
+                    throw new IllegalArgumentException(String.format("Row %s detected with incorrect number of values! Ensure comma-separated value is present for all columns: %s",
```
Indeed! CSVs can be tricky when parsing them manually...although I think I just copied this from the existing parsing logic for OSM 🙈

mark-idleman commented:
Just moved to "ready to review" after adding some examples from manual closure testing I completed with a local router! For the testing, I included a no-closure-applied car routing profile alongside the profile with closures applied to show the contrast between closure/no closure routes (in case it's confusing to see route options go over the closed road sections)

---

@sudatta-mohanty , thanks for taking a look! re: your question about how we might transition to an SEID-only world in the future, I agree that some slightly clunky logic was required at several spots to check whether we're working with OSM or SEIDs, but given that we chose to support both during the transition period, I haven't thought of a cleaner way to accomplish this that would make teardown of OSM logic in the future any cleaner.

As far as I can tell, we'll need to have a way to store both types of IDs internally during the import process and apply different logic in certain places based on the type of IDs present (but I'm open to ideas if there's a better way!).

Still, I think if we did decide to teardown OSM in the future, it would be relatively straightforward. We'd need to get rid of the OSM-specific ID + speed classes and the parent classes I introduced (`CustomSpeed` and `CustomSpeedId`) - the latter of which is mostly a refactor - and then remove the several spots where OSM-specific custom speed logic is applied.

sudatta-mohanty reviewed (approved):
Thanks for the response to the comment! LGTM!

kaelgreco reviewed (approved):
lgtm

Thanks for the comprehensive writeup and testing!

We don't do any scenario-esque testing in the functional test, right?

mark-idleman commented:
> We don't do any scenario-esque testing in the functional test, right?

Correct! The only scenario-like tests we have are in unit tests right now. It could definitely be interesting as a follow-up to set up a new FT that tests specific scenario closures and verifies the results
